### DISCUSSION TITLE
Plugin API for Chalet?

### DISCUSSION THREAD

**Discussion by Prox501** [chalet-org/chalet#155] [Ideas] [answered]
Hello!

I really like how Chalet is designed and I have been wondering if it is possible for Chalet to provide a plugin API to not only C++, but other languages as well so we can extend Chalet for added functionalities like generating Github Action Workflows for a given project, adding custom compilation stages/optimization (such as compiling to LLVM-IR, running custom LLVM-IR optimizer, and then run LLVM-Link), and so forth. 

Would the idea of adding Plugin API feature be welcome in Chalet project?

**Comment by rewrking** [comment, marked as answer]
At the moment, Chalet is definitely focused on core functionality first, so if it's a major language that's widely used and requires compilation, I'd rather see it in that core functionality. Other major languages could be supported like so:

```json
{
   "setttings:Java": {},
   "settings:Swift": {}
}
```

and then chosen in each target with `"language": "Java"` or `"language": "Swift"`

Additionally, you can currently create [process targets](https://www.chalet-work.space/docs/creating-targets?id=process-targets) to call whatever you want and pass along Chalet variables like the architecture triple. That's probably the closest thing to a Plugin API right now.

With that said, a separate one is not out of the question, but there's just a lot of challenges that would bring. For example, how to deploy and consume a plugin -- that sort of thing would require web architecture that doesn't exist yet, and I think we could do better than "download the plugin somewhere and put it in a directory" (although obviously the app itself is this way right now on Mac/Linux)
